### DISCUSSION TITLE
VSCode Extension Functional/E2E Testing

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#194] [Ideas]
As we're working through the VSCode Extension MVP, I wanted to see what our options are for performing some functional and end-to-end tests on the plugin in our pipeline. It looks like the strategy is fairly simple, launch a dockerized web-based vscode instance with your plugin installed, and exercise it with a tool suite like Selenium or Cypress.

I've found a nice write-up on [Testing VSCode Extensions with Cypress and code-server](https://juanmanuelalloron.com/2020/05/05/testing-vscode-extensions-with-cypress-and-code-server/) by Juan Manuel Allo Ron, where he describes how to leverage docker, vscode server, and Cypress to test a vscode extension like a user would interact with the plugin. 

We should decide between Cypress or Selenium and then attempt to produce a github actions workflow that's able to run a test or two with this proposed setup to demonstrate its feasibility and set the groundwork for more extension functional and end-to-end testing. **My personal preference is to try Cypress out**, it looks slick and provides an easier DSL than the pure JS that Selenium uses. If anyone has any additional suggestions/thoughts I'd love to have the feedback!

### Github Actions + Docker Containers
It looks like we can run docker containers as services in our pipeline [see the "Running multiple docker containers in a job" section](https://dev.to/mihinduranasinghe/using-docker-containers-in-jobs-github-actions-3eof)

### Cypress 
[Cypress](https://www.cypress.io/) is an end-to-end feature testing library for web-based interfaces. It has some great-looking interfaces and is JS/TS based (Selenium also supports JS tests). 

Here is a great overview video of Cypress that I found by [Fireship](https://www.youtube.com/watch?v=7N63cMKosIE) (9 min 33 sec long).

### Selenium
[Selenium WebDriver](https://www.selenium.dev/documentation/webdriver/) is a toolsuite that can manage a web browser testing environment to run end-to-end functional tests on web-based UIs. 

Redhat apparently has a [Selenium Webdriver extension](https://github.com/redhat-developer/vscode-extension-tester) specifically for testing vscode extensions. They also have an article, [New tools for automating end-to-end tests for VS Code extensions
](https://developers.redhat.com/blog/2019/11/18/new-tools-for-automating-end-to-end-tests-for-vs-code-extensions#) that describes how to implement the extension to test VSCode extensions, and it looks pretty easy.

### Extension Bundling
There are some requirements to "bundle" VSCode extensions for web environments; we may have to make sure that we're bundling the plugin before trying to test it in this environment. See the documentation about [extension bundling](https://code.visualstudio.com/api/working-with-extensions/bundling-extension).

**Comment by jondavid-black** [comment]
I've heard other teams say nice things about Cypress.  Let me follow up with them and get you some feedback.

**Reply by jondavid-black** [reply]
The feedback I received from the team is they strongly prefer Cypress.  The learning curve is a few days, but once you get things setup it is provided an improved workflow with lower complexity for test development than Selenium.  The cucumber extension is very nice and allows you to just write standard feature files with JS/TS step definitions.  I looked over some examples and they appear pretty clean.  If you'd like I can connect you with the other team.
